### PR TITLE
add anti-spoof-mn3 model to interactive_face_detection_demo

### PR DESCRIPTION
**Pull request by kirillProkofiev**

### CODE REVIEW

**Comment by eaidova** [comment]
Jenkins ok to test

File: demos/interactive_face_detection_demo/detectors.cpp [hunk 1]
```diff
@@ -285,6 +285,93 @@ void FaceDetection::fetchResults() {
...
+
+    matU8ToBlob<uint8_t>(face, inputBlob, enquedFaces);
+
+    enquedFaces++;
+}
+
+float AntispoofingClassificator::operator[] (int idx) const {
+
+    Blob::Ptr  ProbBlob = request->GetBlob(prob_output);
+    auto blobsize = ProbBlob->getTensorDesc().getDims();
```
Comment (Wovchena) [inline demos/interactive_face_detection_demo/detectors.cpp:327]:
```suggestion
```

File: demos/interactive_face_detection_demo/detectors.cpp [hunk 2]
```diff
@@ -285,6 +285,93 @@ void FaceDetection::fetchResults() {
...
+
+    Blob::Ptr  inputBlob = request->GetBlob(input);
+
+    matU8ToBlob<uint8_t>(face, inputBlob, enquedFaces);
+
+    enquedFaces++;
+}
+
+float AntispoofingClassificator::operator[] (int idx) const {
+
```
Comment (Wovchena) [inline demos/interactive_face_detection_demo/detectors.cpp:325]:
```suggestion
```

File: demos/interactive_face_detection_demo/detectors.cpp [hunk 3]
```diff
@@ -285,6 +285,93 @@ void FaceDetection::fetchResults() {
...
+    Blob::Ptr  ProbBlob = request->GetBlob(prob_output);
+    auto blobsize = ProbBlob->getTensorDesc().getDims();
+    LockedMemory<const void> ProbBlobMapped = as<MemoryBlob>(ProbBlob)->rmap();
+    // use prediction for real face only
+    float r = ProbBlobMapped.as<float*>()[2*idx] * 100;
```
Comment (Wovchena) [inline demos/interactive_face_detection_demo/detectors.cpp:330]:
```suggestion
    float r = ProbBlobMapped.as<float*>()[2 * idx] * 100;
```

File: demos/interactive_face_detection_demo/detectors.cpp [hunk 4]
```diff
@@ -285,6 +285,93 @@ void FaceDetection::fetchResults() {
...
+    if (doRawOutputMessages) {
+        std::cout << "[" << idx << "] element, real face probability = " << r << std::endl;
+
```
Comment (Wovchena) [inline demos/interactive_face_detection_demo/detectors.cpp:333]:
```suggestion
```

File: demos/interactive_face_detection_demo/face.cpp [hunk 5]
```diff
@@ -57,6 +62,10 @@ bool Face::isMale() {
     return _maleScore > _femaleScore;
 }
 
+bool Face::isSpoof() {
+    return _spoof_confidence <= 50.;
```
Comment (Wovchena) [inline demos/interactive_face_detection_demo/face.cpp:66]: If net's confidence that a face is spoof is low, then the face should be treated is real. You should name the method `isReal()` or negate the return value.

**Review by Wovchena — CHANGES_REQUESTED** [review]
If `anti-spoof-mn3` isn't going to be added in this PR, you should comment out an update for `models.lst` so it could be readded later.

File: demos/interactive_face_detection_demo/visualizer.cpp [hunk 6]
```diff
@@ -188,16 +188,23 @@ void Visualizer::enableEmotionBar(std::vector<std::string> const& emotionNames)
 
 void Visualizer::drawFace(cv::Mat& img, Face::Ptr f, bool drawEmotionBar) {
     auto genderColor = (f->isAgeGenderEnabled()) ?
-                       ((f->isMale()) ? cv::Scalar(255, 0, 0) :
-                                        cv::Scalar(147, 20, 255)) :
-                                        cv::Scalar(100, 100, 100);
+        ((f->isMale()) ? cv::Scalar(255, 0, 0) :
+            cv::Scalar(147, 20, 255)) :
+            cv::Scalar(192, 192, 192);
```
[the remark this answers is not in the document]
Comment (kirillProkofiev) [reply demos/interactive_face_detection_demo/visualizer.cpp:193]: The text is very hard to see with this color. I changed it to gray.

**Comment by vladimir-dudnik** [comment]
please fix pre-commit check issues

11:26:24  /home/jenkins/workspace/omz/precommit/demos_ubuntu18/src/demos/interactive_face_detection_demo/face.hpp:63:32: error: 'Face::_headPose' will be initialized after [-Werror=reorder]
11:26:24       HeadPoseDetection::Results _headPose;

File: demos/interactive_face_detection_demo/main.cpp [hunk 7]
```diff
@@ -222,6 +227,7 @@ int main(int argc, char *argv[]) {
                 headPoseDetector.submitRequest();
                 emotionsDetector.submitRequest();
                 facialLandmarksDetector.submitRequest();
+                antispoofingClassificator.submitRequest();
```
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/interactive_face_detection_demo/main.cpp]: ping

File: demos/interactive_face_detection_demo/visualizer.cpp [hunk 8]
```diff
@@ -188,16 +188,23 @@ void Visualizer::enableEmotionBar(std::vector<std::string> const& emotionNames)
...
     if (f->isAgeGenderEnabled()) {
         out << (f->isMale() ? "Male" : "Female");
         out << "," << f->getAge();
     }
 
+    if (f->isAntispoofingEnabled()) {
+        float real_confidence = f->getSpoofConfidence();
+        bool SpoofLabel = f->isSpoof();
+        (SpoofLabel) ? out << ",spoof:" << std::setprecision(4) << 100. - real_confidence << '%' :
+            out << ",real:" << std::setprecision(4) << real_confidence << '%';
```
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/interactive_face_detection_demo/visualizer.cpp]: Remove the brackets around `f->isReal()` or put that conditional operator after `out <<`.

**Review by Wovchena — APPROVED** [review]

File: demos/interactive_face_detection_demo/interactive_face_detection.hpp [hunk 9]
```diff
@@ -21,38 +21,45 @@ static const char age_gender_model_message[] = "Optional. Path to an .xml file w
 static const char head_pose_model_message[] = "Optional. Path to an .xml file with a trained Head Pose Estimation model.";
 static const char emotions_model_message[] = "Optional. Path to an .xml file with a trained Emotions Recognition model.";
 static const char facial_landmarks_model_message[] = "Optional. Path to an .xml file with a trained Facial Landmarks Estimation model.";
+static const char antispoofing_model_message[] = "Optional. Path to an .xml file with a trained Antispoofing Classification model.";
 static const char target_device_message[] = "Optional. Target device for Face Detection network (the list of available devices is shown below). "
-                                            "Default value is CPU. Use \"-d HETERO:<comma-separated_devices_list>\" format to specify HETERO plugin. "
-                                            "The demo will look for a suitable plugin for a specified device.";
+"Default value is CPU. Use \"-d HETERO:<comma-separated_devices_list>\" format to specify HETERO plugin. "
+"The demo will look for a suitable plugin for a specified device.";
```
Comment (IRDonch) [inline demos/interactive_face_detection_demo/interactive_face_detection.hpp:27]: Please restore the alignment.

File: demos/interactive_face_detection_demo/models.lst [hunk 10]
```diff
@@ -5,3 +5,4 @@ face-detection-adas-????
 face-detection-retail-????
 facial-landmarks-35-adas-????
 head-pose-estimation-adas-????
+anti-spoof-mn3
```
Comment (IRDonch) [inline demos/interactive_face_detection_demo/models.lst:8]: Please maintain alphabetical order.

File: demos/interactive_face_detection_demo/main.cpp [hunk 11]
```diff
@@ -84,14 +84,16 @@ int main(int argc, char *argv[]) {
             {FLAGS_d_ag, FLAGS_m_ag},
             {FLAGS_d_hp, FLAGS_m_hp},
             {FLAGS_d_em, FLAGS_m_em},
-            {FLAGS_d_lm, FLAGS_m_lm}
+            {FLAGS_d_lm, FLAGS_m_lm},
+            {FLAGS_d_am, FLAGS_m_am}
```
Comment (IRDonch) [inline demos/interactive_face_detection_demo/main.cpp:88]:
```suggestion
            {FLAGS_d_am, FLAGS_m_am},
```

**Review by IRDonch — CHANGES_REQUESTED** [review]

File: demos/interactive_face_detection_demo/README.md [hunk 12]
```diff
@@ -3,20 +3,21 @@
 This demo showcases Object Detection task applied for face recognition using sequence of neural networks.
 Async API can improve overall frame-rate of the application, because rather than wait for inference to complete,
 the application can continue operating on the host while accelerator is busy.
-This demo executes four parallel infer requests for the Age/Gender Recognition, Head Pose Estimation, Emotions Recognition, and Facial Landmarks Detection networks that run simultaneously. You can use a set of the following pre-trained models with the demo:
+This demo executes five parallel infer requests for the Age/Gender Recognition, Head Pose Estimation, Emotions Recognition, Facial Landmarks Detection and Antispoofing Classificator networks that run simultaneously. You can use a set of the following pre-trained models with the demo:
```
Comment (IRDonch) [inline demos/interactive_face_detection_demo/README.md:6]:
```suggestion
This demo executes five parallel infer requests for the Age/Gender Recognition, Head Pose Estimation, Emotions Recognition, Facial Landmarks Detection and Antispoofing Classifier networks that run simultaneously. You can use a set of the following pre-trained models with the demo:
```

File: demos/interactive_face_detection_demo/interactive_face_detection.hpp (hunk 9 above)
Comment (vladimir-dudnik) [reply demos/interactive_face_detection_demo/interactive_face_detection.hpp]: seems to be done

File: demos/interactive_face_detection_demo/models.lst (hunk 10 above)
Comment (vladimir-dudnik) [reply demos/interactive_face_detection_demo/models.lst]: seems to be done

File: demos/interactive_face_detection_demo/main.cpp (hunk 11 above)
Comment (vladimir-dudnik) [reply demos/interactive_face_detection_demo/main.cpp]: seems to be done

File: demos/interactive_face_detection_demo/README.md (hunk 12 above)
Comment (vladimir-dudnik) [reply demos/interactive_face_detection_demo/README.md]: @kirillProkofiev please provide comments if something was changed, We should not look through your latest commits in order to understand if you addressed raised questions.

Comment (kirillProkofiev) [reply demos/interactive_face_detection_demo/README.md]: next time I will know, thanks.
@IRDonch  All issues were resolved

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

**Review by vladimir-dudnik — APPROVED** [review]

**Review by IRDonch — APPROVED** [review]

**Pull request merged**